server/report.go

```diff
@@ -0,0 +1,397 @@
+				b.WriteString("\n\n**Skipped the rest of the services.**")
+				return b.String()
+			}
+
+			if abnormalOnly && s.state == okState {
+				continue
+			}
+
+			if !hostWritten {
+				hostWritten = true
+
+				if linesWritten > 0 {
+					b.WriteString("\n\n")
+				}
+
+				b.WriteString(fmt.Sprintf("`%s`:", h))
+				linesWritten++
+
+				if linesWritten == maximumReportLength {
+					continue
+				}
+
+				// This additional newline produces better Markdown, but we want
+				// to write it only if theEnd hasn't been written.
+				b.WriteRune('\n')
+			}
+
+			b.WriteString(fmt.Sprintf("\n- %s `%s` %s", emoji(s.state), s.name, strings.ToUpper(s.state)))
+			linesWritten++
+		}
+	}
+
+	if linesWritten == 0 {
+		b.WriteString("No services to show.")
+	}
+
+	return b.String()
+}
+
+func (p *Plugin) sendMessages(channelID string, messages ...string) error {
+	for _, m := range messages {
+		post := &model.Post{
+			UserId:    p.botUserID,
+			ChannelId: channelID,
+			Message:   m,
+		}
+		if _, err := p.API.CreatePost(post); err != nil {
+			return fmt.Errorf("p.API.CreatePost: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (p *Plugin) sendMonitoringReport(channelID string) error {
+	hostCountReq := nagios.HostCountRequest{
+		GeneralHostRequest: nagios.GeneralHostRequest{
+			FormatOptions: nagios.FormatOptions{
+				Enumerate: true,
+			},
+		}}
+
+	var hostCount nagios.HostCount
+
+	if err := p.client.Query(hostCountReq, &hostCount); err != nil {
+		return fmt.Errorf("client.Query: %w", err)
+	}
+
+	hostListReq := nagios.HostListRequest{
+		GeneralHostRequest: nagios.GeneralHostRequest{
+			FormatOptions: nagios.FormatOptions{
+				Enumerate: true,
+			},
+		}}
+
+	var hostList nagios.HostList
+
+	if err := p.client.Query(hostListReq, &hostList); err != nil {
+		return fmt.Errorf("client.Query: %w", err)
+	}
+
+	serviceCountReq := nagios.ServiceCountRequest{
+		GeneralServiceRequest: nagios.GeneralServiceRequest{
+			FormatOptions: nagios.FormatOptions{
+				Enumerate: true,
+			},
+		}}
+
+	var serviceCount nagios.ServiceCount
+
+	if err := p.client.Query(serviceCountReq, &serviceCount); err != nil {
+		return fmt.Errorf("client.Query: %w", err)
+	}
+
+	serviceListReq := nagios.ServiceListRequest{
+		GeneralServiceRequest: nagios.GeneralServiceRequest{
+			FormatOptions: nagios.FormatOptions{
+				Enumerate: true,
+			},
+		}}
+
+	var serviceList nagios.ServiceList
+
+	if err := p.client.Query(serviceListReq, &serviceList); err != nil {
+		return fmt.Errorf("client.Query: %w", err)
+	}
+
+	if err := p.sendMessages(
+		channelID,
+		reportPreamble(time.Now()),
+		formatHostCount(hostCount),
+		formatHostList(hostList),
+		formatServiceCount(serviceCount),
+		formatServiceList(serviceList)); err != nil {
+		return fmt.Errorf("p.sendMessages: %w", err)
+	}
+
+	return nil
+}
+
+const reportLockKey = "report-lock"
+
+var reportLock = []byte{0b10, 0b1, 0b11, 0b111}
+
+func (p *Plugin) monitoringReportLoop() {
+	for {
```

Comment:
This looks like a busy waiting. why not use a cron job instead?

Reply:
I believe that's not enough for the HA environment. If I understand your suggestion correctly, the _cron job_ would be running the reporting loop on the node selected after subscription. I'm afraid that when the plugin instance on that node dies, the system monitoring report will be no longer delivered. Instead, I'm proposing a solution where every node runs the loop, but only one of them enters the critical section. Every node waits a minute (using `time.Sleep`) and tries to enter the critical section to build the report. The lock has a duration of the system monitoring report frequency. Only one node does that since this is a mutually exclusive action. I also believe this is approximately as efficient as a non-HA solution; there's no busy waiting in its classical definition because we use `time.Sleep`.

Reply:
I'd encourage using the scheduling mechanism from [here](https://github.com/mattermost/mattermost-plugin-api/blob/1d382e5ea2077dfd81933aed74222fd5b7e2f8cf/cluster/job.go#L110). It is implemented using the distributed lock and you will be guaranteed to have one `monitoringReportLoop` run in every X seconds in an HA environment.

Reply:
Done. Thank you. It made this part of the code substantially leaner.